src/Meditation.MetadataLoaderService/Services/MetadataLoader.cs

```diff
@@ -4,36 +4,60 @@
 using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
 
 namespace Meditation.MetadataLoaderService.Services
 {
     internal class MetadataLoader : IMetadataLoader
     {
-        private readonly ConcurrentDictionary<string, AssemblyDef> _loadedAssemblies;
+        private readonly ConcurrentDictionary<string, AssemblyDef> _assemblies;
         private readonly ConcurrentDictionary<string, AssemblyMetadataEntry> _metadataModels;
         private readonly object _syncObject;
 
         public MetadataLoader()
         {
-            _loadedAssemblies = new ConcurrentDictionary<string, AssemblyDef>();
+            _assemblies = new ConcurrentDictionary<string, AssemblyDef>();
             _metadataModels = new ConcurrentDictionary<string, AssemblyMetadataEntry>();
             _syncObject = new object();
         }
 
-        public AssemblyMetadataEntry LoadMetadataFromAssembly(string path)
+        public IEnumerable<AssemblyMetadataEntry> LoadMetadataFromProcess(IEnumerable<string> modulePaths)
+        {
+            var assembliesLookup = new Dictionary<(UTF8String Name, Version Version), AssemblyDef>();
+            foreach (var modulePath in modulePaths.OrderBy(Path.GetFileName))
+            {
+                var (metadata, assemblyDef) = LoadMetadataFromAssemblyImpl(modulePath);
+                if (assembliesLookup.ContainsKey((assemblyDef.Name, assemblyDef.Version)))
```

Comment:
[Question] What if we'll get NGEN image first? Is it ok? Do you need only name/version?

Reply:
> What if we'll get NGEN image first? Is it ok?

It is OK. Getting NGEN image first is normal (or in fact getting only native image is normal). They have identical metadata entries so we do not really care which is loaded. Native images are supposed to be just performance optimization. Runtime seems to prefer native images when they are available: https://github.com/dotnet/runtime/blob/5f4ee519c7ce0bfe1e914ee7a617e17b5c509cf5/src/coreclr/binder/utils.cpp#L185

> Do you need only name/version?

According to documentation if we want to implement correct behavior, we should also consider culture and public key (if it is available). See this: https://learn.microsoft.com/en-us/dotnet/framework/deployment/how-the-runtime-locates-assemblies